src/west/app/project.py

```diff
@@ -640,6 +652,68 @@ def _die_if_manifest_project_filter(self, action):
                      'the manifest while projects are made inactive by the '
                      'project filter.')
 
+    def _untracked(self):
+        ''' "Performs a top-down search of the west topdir,
+        ignoring every folder that corresponds to a west project.
+        '''
+        ppaths = []
+        untracked = []
+        for project in self._projects(None):
+            # We do not check for self.manifest.is_active(project) because
+            # inactive projects are still considered "tracked folders".
+            ppaths.append(Path(project.abspath).absolute())
+
+        # Since west tolerates nested projects (i.e. a project inside the folder
+        # of another project) we must sort the project paths to ensure that we
+        # hit the "enclosing" project first when iterating.
+        ppaths.sort()
+
+        def _find_untracked(folder):
+            '''There are three cases for each element in a folder:
+            - It's a project -> Do nothing, ignore the folder.
+            - There are no projects inside -> add to untracked list.
+            - It's not a project folder but there are some projects inside it -> recurse.
+            The folder argument cannot be inside a project, otherwise all bets are off.
+            '''
+            self.dbg(f'looking for untracked files/folders in: {folder}')
+            for e in [e.absolute() for e in folder.iterdir()]:
+                if not e.is_dir():
+                    untracked.append(e)
+                    continue
+                self.dbg(f'processing folder: {e}')
+                for ppath in ppaths:
+                    # We cannot use samefile() because it requires the file
+                    # to exist (not always the case with inactive or even
+                    # uncloned projects).
+                    if ppath == e:
```

Comment:
I wonder whether symbolic links could break this, let me try.
If yes then maybe you should first check whether ppath exists, then compare.

Or maybe even simpler, something like `ppaths = [ p for p in ppaths if p.exists() ]`

Symbolic links can be a nightmare: https://lwn.net/Articles/899543/

But symlinks are also incredibly popular on Linux. I don't think `west` should bend over backwards to support symbolic links BUT it should work reasonably well in simple cases like `/home/me -> /disk5/home/me`.

- #313
- https://github.com/zephyrproject-rtos/zephyr/issues/50284 EVERYTHING is a symlink in NixOS!!

Reply:
@marc-hb  could you explain what you mean by "break this"? i.e. could you provide a way for me to reproduce the potential issue locally for me to understand it better before proposing a solution?

Reply:
> > I wonder whether symbolic links could break this, let me try.

>  i.e. could you provide a way for me to reproduce the potential issue locally for me to understand it better before proposing a solution?

Wait, I don't know whether I can "break" it yet! Please give me a moment.

Reply:
I cannot use this @marc-hb. When projects are inactive (or even uncloned!) the folder is not present on disk, so `p.exists()` would not include those.

Reply:
@carlescufi found a great test case where my suggestion changes the output:

- project `foo/bar/baz` is not cloned and ppath.exists() is false
- User creates  `foo/bar/untracked1`, `foo/bar/untracked2`, ...

Then the output differs.

I think this demonstrates that tests are the "real" specification and the only way to formalize ambiguities.